dll/win32/msgina/gui.c

```diff
@@ -989,8 +991,22 @@ SecurityDialogProc(
                         EndDialog(hwndDlg, WLX_SAS_ACTION_LOGOFF);
                     return TRUE;
                 case IDC_SECURITY_SHUTDOWN:
-                    if (OnShutDown(hwndDlg, pgContext) == IDOK)
+                    /* Emergency restart feature */
+                    if (ISKEYDOWN(VK_CONTROL))
+                    {
+                        if (ResourceMessageBox(pgContext,
+                                               hwndDlg,
+                                               MB_OKCANCEL | MB_ICONSTOP,
+                                               IDS_EMERGENCY_RESTART_TITLE,
+                                               IDS_EMERGENCY_RESTART))
+                        {
+                            NtShutdownSystem(ShutdownReboot);
```

Comment:
Unless MSGINA already has the `SeShutdownPrivilege` privilege assigned (and enabled) in the context of the calling thread then all is good but in case it's not: you are invoking a power action operation that requires this specific privilege in order to restart the system.

You could use `RtlAdjustPrivilege` to enable that privilege if you're impersonating the SYSTEM's security context or the MSGINA process' access token has that privilege.

By the way, since you're directly doing a power action request in MSGINA, I would argue that such kind of thing has to be done in WINLOGON instead. This is because of how MSGINA and WINLOGON interface together. You could achieve that by just calling `EndDialog` with a return value of something like `WLX_SAS_ACTION_EMERGENCY_REBOOT` or whatever and then you handle your power actions there once you receive that WLX message from the message pump. If that sounds too complicated then add a TODO comment with an explanation in regard of what it should be done in the future.

Reply:
> This is because of how MSGINA and WINLOGON interface together. You could achieve that by just calling EndDialog with a return value of something like WLX_SAS_ACTION_EMERGENCY_REBOOT or whatever and then you handle your power actions there once you receive that WLX message from the message pump.

Well, this is not what Windows does.

Reply:
Ok correction, I tested this feature on my W2k3 debug VM just out of curiosity and apparently both the modal dialog and MSGINA dialog aren't killed which implies the power action is done entirely in MSGINA. My bad.